### DISCUSSION TITLE
Symmetric Cartesian Simulation

### DISCUSSION THREAD

**Discussion by gsgall** [idaholab/moose#25611] [Q&A Modules: Navier-Stokes] [answered]
I am trying to simulate flow from a square channel onto a surface. To start I am only interested in the velocity and pressure profiles but later I will build on this to simulate some gas mixing. I am currently using a 2D plane in the middle of my pipe and simulating both sides of the flow and the results look good but take a very long time to converge in the region I am interested in. I would like to be able to cut the domain in half and only simulate half of this cross section and I would like to do so while solving in cartesian coordinates.

In my mesh I have a boundary called `axis_of_symmetry` and this is in the center of the channel. On this boundary I have tried `INSADMomentumNoBCBC` and a natural boundary condition. With either boundary condition the velocity at the center of this boundary becomes non-physical very quickly 

![Screenshot 2023-09-27 at 10 55 15 PM](https://github.com/idaholab/moose/assets/66559200/77ebb7c2-5690-494c-93ce-5c73402da492)

However, when I put this simulation in cartesian coordinates I have no issues with this boundary.  

My input is below

```
[GlobalParams]
  integrate_p_by_parts = false
[]

[Mesh]
  file = 'cost_jet_half.msh'
  second_order = true
[]

[Variables]
  [velocity]
    order = SECOND
    family = LAGRANGE_VEC
    block = 'plasma'
  []

  [p]
    order = SECOND
    family = LAGRANGE
    block = 'plasma'
  []
[]

[Kernels]
  [mass]
    type = INSADMass
    variable = p
    block = 'plasma'
  []

  [momentum_time_derivative]
    type = INSADMomentumTimeDerivative
    variable = velocity
    block = 'plasma'
  []

  [momentum_convection]
    type = INSADMomentumAdvection
    variable = velocity
    block = 'plasma'
  []

  [momentum_viscous]
    type = INSADMomentumViscous
    variable = velocity
    block = 'plasma'
  []

  [momentum_pressure]
    type = INSADMomentumPressure
    variable = velocity
    pressure = p
    block = 'plasma'
  []

  [supg]
    type = INSADMomentumSUPG
    variable = velocity
    velocity = velocity
    block = 'plasma'
  []
  # [gravity]
  #   type = INSADGravityForce
  #   variable = velocity
  #   gravity = '0 -9.81 0'
  # []
[]

[AuxVariables]
  [vel_x]
    order = SECOND
    block = 'plasma'
  []
  [vel_y]
    order = SECOND
    block = 'plasma'
  []
[]

[AuxKernels]
  [vel_x]
    type = VectorVariableComponentAux
    variable = vel_x
    vector_variable = velocity
    component = 'x'
  []
  [vel_y]
    type = VectorVariableComponentAux
    variable = vel_y
    vector_variable = velocity
    component = 'y'
  []
[]

[ICs]
  [pressure]
    type = ConstantIC
    variable = p
    value = 0
    block = 'plasma'
  []

  [velocity]
    type = VectorFunctionIC
    variable = velocity
    function_x = 0
    function_y = channel_func
    block = 'plasma'
  []
[]
[BCs]
  [inlet]
    type = VectorFunctionDirichletBC
    variable = velocity
    boundary = 'inlet'
    function_x = 0
    function_y = 'inlet_func'
  []

  [wall]
    type = VectorFunctionDirichletBC
    variable = velocity
    boundary = 'electrode target upper_atmosphere atmosphere'
    function_x = 0
    function_y = 0
  []

  [pressure_condition]
    type = PenaltyDirichletBC
    variable = p
    boundary = 'atmosphere upper_atmosphere'
    value = 101325
    penalty = 1e5
  []
[]

[Functions]
  # convert flow rate to m^3/s
  [flow_rate_m3_s]
    type = ParsedFunction
    # flow rate in slm
    # other dimensions in m
    symbol_names = 'flow_rate mins_to_sec l_to_m3'
    symbol_values = '1.0      60          1e3'
    expression = 'flow_rate / (l_to_m3 * mins_to_sec)'
  []
  # converting volumetric flow rate to velocity using
  # cross sectional area
  [max_vel]
    type = ParsedFunction
    symbol_names = 'flow_rate       channel_width channel_depth'
    symbol_values = 'flow_rate_m3_s 1e-3          1e-3'
    expression = 'flow_rate / (channel_width * channel_depth)'
  []

  [inlet_r_start]
    type = ParsedFunction
    value = '-0.5 / 1000'
  []

  [inlet_r_end]
    type = ParsedFunction
    value = '0.5 / 1000'
  []

  [inlet_r_center]
    type = ParsedFunction
    vars = 'inlet_r_start inlet_r_end'
    vals = 'inlet_r_start inlet_r_end'
    value = '( inlet_r_start + inlet_r_end ) / 2'
  []

  [inlet_func]
    type = ParsedFunction
    vars = 'inlet_r_start inlet_r_end inlet_r_center max_vel'
    vals = 'inlet_r_start inlet_r_end inlet_r_center max_vel'
    value = '-max_vel * ( ( x - inlet_r_start ) * ( x - inlet_r_end ) / ( ( inlet_r_center - inlet_r_start ) * ( inlet_r_center - inlet_r_end ) ) )'
  []

  [channel_func]
    type = ParsedFunction
    symbol_names = 'inlet_func'
    symbol_values = 'inlet_func'
    expression = 'if (x > -0.5e-3 & x < 0.5e-3,
                  inlet_func,
                  0)'
  []
[]

[Materials]
  [fluid_mats]
    type = ADGenericConstantMaterial
    prop_names = 'rho mu cp k'
    prop_values = '0.1598 1.9e-5 1 1'
    block = 'plasma'
  []

  # [ins_mat]
  #   type = INSADMaterial
  #   velocity = velocity
  #   pressure = p
  # []
  #   block = 'plasma'

  [ins_mat]
    type = INSADTauMaterial
    velocity = velocity
    pressure = p
    block = 'plasma'
    alpha = 1
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
    solve_type = 'NEWTON'
  []
[]

# [Executioner]
#   type = Steady
#   petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount'
#   petsc_options_value = 'lu NONZERO 1.e-10'
#   # nl_rel_tol = 1e-7
#   nl_max_its = 100
#   automatic_scaling = true
#   compute_scaling_once = false
#   # line_search = none
# []

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount'
  petsc_options_value = 'lu NONZERO 1.e-10'
  line_search = none

  nl_abs_tol = 2e-8
  nl_max_its = 15

  l_max_its = 300
  [TimeStepper]
    type = IterationAdaptiveDT
    cutback_factor = 0.4
    dt = 1e-8
    growth_factor = 1.2
    optimal_iterations = 10
  []
  steady_state_detection = true
  # steady_state_tolerance = 1e-3
  automatic_scaling = true
  compute_scaling_once = false
[]
  
[Outputs]
  console = true
  [out]
    type = Exodus
  []

  [out_nl]
    type = Exodus
    execute_on = 'NONLINEAR'
  []

  [out_l]
    type = Exodus
    execute_on = 'LINEAR'
  []
[]

```

**Comment by gsgall** [comment]
In my discussion there is an error when I wrote

"However, when I put this simulation in cartesian coordinates I have no issues with this boundary."

It should have said 

"However, when I put this simulation in cylindrical coordinates I have no issues with this boundary.

**Comment by GiudGiud** [comment]
So 2D XY is the problem?
And 2D RZ is fine?

**Reply by gsgall** [reply]
Yes that is correct.

**Reply by GiudGiud** [reply]
in which configuration do you get the expected results?
and did you take a look at some RZ inputs for INSAD in the test folder?

**Reply by GiudGiud** [reply]
why do you want to use 2D XY?  Is the system truly XY or it's just a test?

**Reply by gsgall** [reply]
The system is truly XY. The flow channel in the system is a rectangular prism.

**Reply by GiudGiud** [reply]
It must be 3D then right, because XY would be an infinite prism

**Reply by gsgall** [reply]
Yeah, the system is 3D but not axisymmetric so to model the flow from the system. I thought it would be most appropriate to model a cross section of the flow in 2D XY.

**Reply by GiudGiud** [reply]
You wont get the wall effects right  as effectively if you model a cross section you model an infinite prism.
However it may be good enough to get started. 

So what are all the boundary conditions on the pipe part?
In the center it should be natural as you have symmetry right
At the top do you have dirichlet BCs for both components?
and on the wall what do you have?

**Reply by gsgall** [reply]
Since I am mostly interested in the effects of the fluid when it hits a surface would it be more effective to model the system as a cylindrical tube while maintaining the same flow volume in RZ that exists in XYZ

On the walls I have a 0 velocity condition.

**Reply by GiudGiud** [reply]
That should be fine to get started. 

How do you do the outlet? Where is the fluid supposed to exit?
I see 0 dirichlet BCs on a bunch of surface.

**Reply by gsgall** [reply]
The boundary called `inlet` has a half parabola as the inlet condition. It is on the top left corner of the mesh.

**Reply by GiudGiud** [reply]
But what s the outlet

**Reply by gsgall** [reply]
Sorry for my mistake. 

Right now I do not have an outlet condition I am assuming that the system is in a closed box at atmospheric pressure for this simulation.

**Reply by GiudGiud** [reply]
That s going to be problematic. 
INS is for incompressible flow, so adding mass continuously leads to convergence issues. The conservation of mass equation just gets impossible to solve (because mass is not conserved). 

Can you relax one of the dirichlet no-slip BCs and make that an outlet?

**Reply by gsgall** [reply]
I have adjusted the mesh to put the system in 2D RZ and add have added 2 outlet conditions. The new outlet conditions are on the right side of the domain, and those have been set to `INSADMomentumNoBCBC`. I am still getting similar instabilities in the inlet channel and  I am getting an instability at the point where the no-slip condition meets the no BCBC. 
![Screenshot 2023-09-29 at 1 01 34 PM](https://github.com/idaholab/moose/assets/66559200/9601c325-2518-43ed-ab27-4d86cfa5973a)

The corner instability is kind of hard to see so here is a zoom in of it as well 
![Screenshot 2023-09-29 at 1 07 21 PM](https://github.com/idaholab/moose/assets/66559200/9efce491-64fd-4d0b-80e2-517840251cf3)

**Reply by GiudGiud** [reply]
What is the Reynolds number?
@lindsayad do we need a PSPG term in the mass equation?
For the boundary condition, which one takes over at junction?

**Reply by gsgall** [reply]
The Reynolds number is pretty small ~283


But I believe I have uncovered the issue with the inlet channel and the corner instability. Both were caused by a meshing issue. The inlet was not the correct width, I missed a factor of 1/2 in my conversion and the corner which I rounded was assigned to the incorrect boundary name. 

The channel seems to be giving reasonable results now but the boundaries where there are `INSADMomentumNOBCBC` conditions seem to be gaining a small velocity which is unexpected. 

![Screenshot 2023-09-29 at 1 55 36 PM](https://github.com/idaholab/moose/assets/66559200/93bd4e32-bdf3-41f7-a0d7-e799674fb20f)

**Reply by GiudGiud** [reply]
Interesting. Looks like they are becoming inlets.
Can you keep only one outlet?

**Reply by lindsayad** [reply]
The very first thing you should do is either reduce the order of your pressure variable to FIRST, or add PSPG. I would recommend the former if you are going to go with second order for velocity. Equal order for velocity and pressure is unstable

**Reply by lindsayad** [reply]
Also I'm pretty surprised that your LU factorization is working since as far as I can tell you should have a pressure null-space. You are not integrating the pressure by parts which means you only have gradients of pressure in your discretization and so without some other means to pin the pressure value your pressure field should only be defined up to a constant

**Reply by gsgall** [reply]
I set one outlet and the pressure to order FIRST and I am getting similar instabilities on the single outlet. 

I am relatively new to CFD simulations and was advised by another student to set `integrate_by_parts` to false since it would be more accurate. I am also using LU for the solve since we use these options as standard for the plasma simulations I am more accustomed to.  What solve type would be more appropriate for this type of CFD work?

**Reply by GiudGiud** [reply]
For INSAD the Jacobian is numerically acrruate (thanks to AD) so you should be able to use Newton for the solve type. 

My guess is that the artefacts are due to modeling issues. 
As Alex mentioned, we need to remove the pressure null space.Could you pin the pressure at the outlet or at another location in the system?

**Reply by gsgall** [reply]
Adding the pressure pin worked for a case with a single outlet on the side of the domain I was applied the method in `modules/navier_stokes/test/tests/finite_element/ins/RZ_cone/rz_cone_by_parts_steady_nobcbc.i` to my simulation. 

However, when I tried to add multiple outlets to the domain, the side and top of the domain the same kind of instabilities with velocity on the boundaries arise. Is there a recommended method for resolving instabilities for multiple outlets?  Or would it best to try and limit simulations to a single outlet?

**Reply by GiudGiud** [reply]
Setting multiple outlets is fine, but you have to know the pressure at multiple locations, and that's not always obvious. It's easy to create an unphysical situation. 

If you set an outlet at P1, and another one at P2 > P1, if you flow coming from the inlet is not such that these two conditions can be met by flow from the inlet to the two outlets, you may start observing flow from the outlet at P2 to the outlet at P1. 

Here you are setting a large region at a fixed pressure at the top of the system, when the flow is actually mostly parallel to this boundary rather than orthogonal to it (what you expect for a constant boundary). Especially as you do not consider gravity, we kind of expect a horizontal pressure gradient along this top boundary. 
Can you plot the streamlines of the velocity in paraview to check what this actually looks like? I'm speculating here

**Reply by gsgall** [reply]
Yes you are correct it looks like the flow is mostly parallel to the upper boundary.

![Screenshot 2023-09-30 at 11 57 08 AM](https://github.com/idaholab/moose/assets/66559200/cf3246ea-7b89-42c2-9ce3-e9feacc180a2)

**Reply by GiudGiud** [reply]
The ~~left~~ right boundary looks like it could be a good outlet?

The top one probably doesn't not have an outlet in that case?

**Reply by gsgall** [reply]
In that case the top is set to no-slip conditions and the side has a pinned pressure and NoBCBC.

**Reply by GiudGiud** [reply]
so are there issues with the results for that case?
which boundaries would you like to modify to be another outlet?

**Reply by GiudGiud** [reply]
@lindsayad which BC is the outlet without a specified pressure for INSAD?
i m thinking we ll want to pin the volume somewhere in the domain instead of this case with multiple outlets

**Reply by gsgall** [reply]
Ideally I would like to modify the top to be an outlet as well. But those results should be sufficient for the work I am doing as well.

**Reply by lindsayad** [reply]
Oh I missed the penalty Dirichlet condition that you are imposing on the pressure. That does remove the nullspace but it could also perturb the conservation of mass equation

**Reply by lindsayad** [reply]
Can you create a git branch with your input file? It is hard on a longish thread to keep track of the current state of the input file.

In a real experimental setup you would not have an inlet without an outlet. Do you have a specific experimental setup you want to model? Is this a @skatersaurus research project?

**Reply by lindsayad** [reply]
Oh it looks like you are trying to model the COST jet... It's unclear to me how the gas is supposed to flow in that system. @cticenhour are you better acquainted with the COST jet?

**Reply by cticenhour** [reply]
I am not familiar with COST, but @csdechant might be.

**Reply by lindsayad** [reply]
> @lindsayad which BC is the outlet without a specified pressure for INSAD?

If you do not integrate pressure by parts, then no boundary condition will impact the pressure field. If you do integrate by parts, then a natural boundary condition would set the vector condition: n * sigma + n*p = 0 where n is the normal. In a no BC BC, in the continuous setting you are setting no boundary condition, but in the discrete setting I believe you are setting some higher order derivative to 0. Yes, according to [Griffiths](https://www.scopus.com/record/display.uri?eid=2-s2.0-0031080083&origin=inward) with a NoBCBC you are setting the p+1 derivative (where p is the polynomial order of your basis) close to 0 at some point near the outlet which Griffiths showed perturbed upstream conditions less, introducing less error, than a natural boundary condition. (But this would be a condition on the velocity variable, not the pressure variable)

**Reply by gsgall** [reply]
The current working version of the input I am using is here [https://github.com/gsgall/zapdos/tree/marias_jet/jet_fluid_only](url). Results are stable when setting the upper part of the domain to a no-slip condition and pinning the pressure at a node near the top right corner. I would like to be able to make the boundary named `upper_atmosphere` a NoBCBC but when doing so the velocity field becomes unstable near the pinned pressure node. I have tried moving the node near the upper right of the side outlet and the bottom right of the side outlet but similar behavior occurs in both where the solution in the corner begins to create an inlet near the pinned node. The image shows an example when I tried to set the `upper_atmosphere` to a NoBCBC and set the pinned pressure node near the bottom right of the domain.


<img width="861" alt="Screenshot 2023-10-02 at 2 22 57 PM" src="https://github.com/idaholab/moose/assets/66559200/46742a9b-26e9-4137-ba02-c9562684a32c">

**Reply by lindsayad** [reply]
I would try looking at the `ad_lid_driven_mean_zero_pressure.i` test case in `modules/navier_stokes/test/tests/finite_elements/ins/lid_driven`. You can add a few objects, a new variable
```
  [lambda]
    family = SCALAR
    order = FIRST
  []
```
 a kernel
```
 [mean_zero_pressure]
    type = ScalarLagrangeMultiplier
    variable = p
    lambda = lambda
  []
```
scalar kernel
```
[ScalarKernels]
  [mean_zero_pressure_lm]
    type = AverageValueConstraint
    variable = lambda
    pp_name = pressure_integral
    value = 0
  []
[]
```
postprocessor
```
[Postprocessors]
  [pressure_integral]
    type = ElementIntegralVariablePostprocessor
    variable = p
    execute_on = linear
  []
[]
```
and then you can remove the nodal pressure pin. Then we can see whether that removes the artifacts around the pin.

Ultimately, I still think we should identify a true outlet in the COST jet

**Reply by gsgall** [reply]
I will give that a try. 

As for the outlet. The outlet of the jet itself exists near the left boundary. In this simulation I am modeling flow from the jet onto a target and the boundaries that are on the upper and lower right side of the domain need to represent areas that are open to atmosphere.

**Reply by gsgall** [reply]
The mean-zero pressure method worked for my model!

Are there any resources you would recommend so I understand this method a bit better?

**Reply by lindsayad** [reply]
That's great to hear! Yes, for the mean zero treatment you can look at [this pdf](https://github.com/idaholab/large_media/blob/master/framework/scalar_constraint_kernel.pdf)

**Reply by lindsayad** [reply]
As an aside, I'm very glad to see another plasma person working with zapdos/moose!

**Reply by gsgall** [reply]
In the results for the pressure the values are about between 0 and 120 [Pa] which does not correspond to the physical system. But the gradient between the highest and lowest pressures look about the same as when using the pinned pressure method. I am not currently interested in the pressure as much as I am the velocity but if I were would I have to solve this field with something like the pinned pressure method or could I just added the approximate ambient pressure to the result that simulation gives.

I am very excited to work with MOOSE and Zapdos. I think it's a very exciting tool!

**Reply by lindsayad** [reply]
The fact that no part of the continuous (or discrete) equations involves the pressure is the very reason that something like a pressure pinning or mean value treatment is necessary. If you start introducing equations of state that involve the pressure, then you will no longer need something like a pressure pin or the mean treatment.

If you want, you can change the `value` parameter in the `AverageValueConstraint` to change what you want the mean value of the pressure to be. But this is purely for aesthetic purposes. It should not change the velocity field or the pressure gradient in any way

**Reply by gsgall** [reply]
That makes sense, thank you to everyone who helped. I really appreciate it!

**Reply by gsgall** [reply, marked as answer]
The follow steps were taken to fix this issue

1. Approximate the rectangular prism as a cylindrical system, while ensuring the inlet channel volume is maintained. 
2. The order of pressure and velocity were offset so that velocity is second and pressure is first since having them the same order is unstable. 
3. The mean zero pressure method was used to remove the pressure null space from the INS solve. For an example see `modules/navier_stokes/test/tests/finite_element/ins/RZ_cone/ad_lid_driven_mean_zero_pressure.i`

The working input file and .geo files are available here [here](https://github.com/gsgall/zapdos/tree/marias_jet/jet_fluid_only)

**Reply by lindsayad** [reply]
great!
